docs/madr/decisions/070-resource-identifier.md

```diff
@@ -0,0 +1,378 @@
+
+* `name` and `zone`
+  * contain no more than 253 characters
+  * contain only lowercase alphanumeric characters, '-' or '.'
+  * start with an alphanumeric character
+  * end with an alphanumeric character
+* `mesh`
+  * contain no more than 63 characters
+  * contain only lowercase alphanumeric characters, '-' or '.'
+  * start with an alphanumeric character
+  * end with an alphanumeric character
+* `namespace`
+  * contain at most 63 characters
+  * contain only lowercase alphanumeric characters or '-'
+  * start with an alphanumeric character
+  * end with an alphanumeric character
+* `resourceType`
+  * only alphanumeric
+* `sectionName` (even though [ContainerPort's name](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#containerport-v1-core) is limited by 15 characters, [ServicePort's name](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#serviceport-v1-core) can contain upto 63 character)
+  * contain at most 63 characters
+  * contain only lowercase alphanumeric characters or '-'
+  * start with an alphanumeric character
+  * end with an alphanumeric character
+
+This leaves us with the following charset:
+
+```
+delimiter = "_" / "~" / "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "=" / "@"
+```
+
+_Note: if we'd like to use resource identifier in URL query the delimiter charset is significantly smaller:_
+
+```
+delimiter = "_" / "~" 
+```
+
+## Decision Drivers
+
+- New Inspect API endpoints that'd allow getting resolved configuration by the resource identifier
+- Kuma API GET endpoints shouldn't require core name to read the resource 
+- Human-readable: users should be able to type the identifier manually if needed, so it's not hashed
+- Envoy resources that have direct correlation with Kuma resources should be named by using resource identifier, i.e. outbound cluster can be named after MeshService
+- Ability to use resource identifier in Prometheus queries, i.e
+  - `my_metric{envoy_cluster_name~="kri.*_meshservice_.*"} // Get only actual outbounds from meshservice`
+  - `my_metric{envoy_cluster_name~="kri_default_.*.*"} // Get only stuff on the default mesh`
+
+## Considered Options
+
+* Option 1 - Order-based, no field names in the identifier
+* Option 2 - Field names in the identifier
+
+## Decision Outcome
+
+* Option 1 - Order-based, no field names in the identifier
+
+## Pros and Cons of the Options
+
+### Option 1 - Order-based, no field names in the identifier
+
+We need to pick a delimiter that's present in the `delimiter` charset we've specified above.
+
+There is an identifier format from Amazon called [ARN](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference-arns.html). We can adopt a similar approach, but using `_`:
+
+```
+kri_<resource-type>_<mesh>_<zone>_<namespace>_<resource-name>_<section-name>
```

Comment:
Could we have an optional `suffix` ? I already envision use cases where actually we can't sumup with a single kri even with a sectionName and we still need to append some `hash` at the end?

I'd argue strongly against the `hashes` if it came up but I'm sure it's gonna come up.

Another probably better option is to explicitly state that if needs for a suffix was required we will need to create a MADR. This enable us to leave this door open but clearly say it should never be required. From looking at the discussion in: https://github.com/kumahq/kuma/issues/13129 clearly it seems that it'd be necessary

Reply:
This reminded me we should probably treat `sectionName` as an optional suffix, because when it's missing, we don't want to have a trailing `_`, like 
```
kri_meshservice_mesh-1_us-east-1_kuma-demo_mt1_
```

So maybe we can put `sectionName` after some delimiter, because strictly speaking `sectionName` is not a resource identifier, it's a reference inside the specific resource:
```
kri_<resource-type>_<mesh>_<zone>_<namespace>_<resource-name>&sectionName=httpport
OR
kri_<resource-type>_<mesh>_<zone>_<namespace>_<resource-name>&sn=httpport
```

and then like you said if we need another suffix, we will create MADR to add another KV pair behind `&`, i.e. `hash` 
```
kri_<resource-type>_<mesh>_<zone>_<namespace>_<resource-name>&sn=httpport,hash=e3a0069bda479bd1
```

WDYT?

Reply:
I'd rather have section be the default with a trailing `_` the rest can wait for a future MADR

Reply:
Added a note:

> #### ResourceIdentifier format is open for extension in the future MADRs
>
> Today when generating clusters for Builtin Gateway, we use the hash suffix to support a MeshService cluster per MeshGateway listener. If we want cluster name to follow `kri` format, we need to extend the format with a new MADR.